### DISCUSSION TITLE
Data Import QA Summary

### DISCUSSION THREAD

**Discussion by KingSkyLi** [TuGraph-family/tugraph-db#107] [Q&A]
本贴是数据导入的QA汇总贴，详见下方回答。
This post is a QA summary post of data import, see the answer below for details.

**Comment by KingSkyLi** [comment]
中文：
Q: TuGraph 可以对接那些常用数据库？
A: TuGraph通过DataX可以实现大部分主流数据库的导入导出，支持的数据库包括MySQL、Oracle、Hive 等。具体参考https://github.com/TuGraph-db/DataX


English：
Q: What common databases can TuGraph connect to?
A: TuGraph can import and export most mainstream databases through DataX. The supported databases include MySQL, Oracle, Hive, etc. For details, refer to https://github.com/TuGraph-db/DataX

**Comment by KingSkyLi** [comment]
中文：
Q：如何进行数据导入？
A：数据导入可见官网文档：[https://www.tugraph.org/doc?version=V3.3.0&id=10000000000658643](https://www.tugraph.org/doc?version=V3.3.0&id=10000000000658643)

English：
Q: How to import data?
A: Data import can be found in the official website document: [https://www.tugraph.org/doc?version=V3.3.0&id=10000000000658643](https://www.tugraph.org/doc?version=V3.3.0&id=10000000000658643)

**Comment by zhangwh807** [comment]
中文：
Q：读取oracle数据报错
"error_message":"Error parsing file memory_file_stream\n\tError occurred at offset 0, exception detail:\n\tjson reading failed, error msg : std::bad_cast\n>Error line...."，如何解决？
A：看起来像在处理数据的时候遇到特使符号导致报错的，建议用相对较小的表以及数据可以尝试测一下

English：
Q: What can I do if the following error occurs when reading oracle datagram？
"error_message":"Error parsing file memory_file_stream\n\tError occurred at offset 0, exception detail:\n\tjson reading failed, error msg : std::bad_cast\n>Error line...."
A:  It seems that the envoy symbol caused an error when processing data, it is recommended to use a relatively small table and the data can try to test it

**Comment by zhangwh807** [comment]
中文：
Q：value pack时的null array的具体含义是？
A：标记这个schema的field是否为空，如果一个schema的field为空，并且插入的数据里对应的列是空的，在packed的时候就不占内存

English：
Q: What is the meaning of "null array" in "value pack"?
A: Mark whether the field of the schema is empty. If the field of a schema is empty and the corresponding column in the inserted data is empty, it does not take up memory when packed

**Comment by zhangwh807** [comment]
中文：
Q：怎么把存储于tugraph的某些指定的点/边类型的全量数据，导出到csv文件中？
A：使用neo4j driver 连接tugraph，直接发送cypher 语句 "match (n) return n" 就可以了。结果是流式返回的，不管多少数据都可以读出来，不会引发内存oom

English：
Q: How to export the full amount of data stored in tugraph for certain specified vertex/edge types to a csv file?
A: Use the neo4j driver to connect to tugraph and simply send the cypher statement "match (n) return n". The result is streaming, no matter how much data can be read, without triggering memory oom
